File: 0084-Largest-Rectangle-in-Histogram/solution.py
class Solution:
    def largestRectangleArea(self, heights: list[int]) -> int:
        stack = []  # stores indices
        max_area = 0
        n = len(heights)

        for i in range(n + 1):
            curr_height = 0 if i == n else heights[i]

            while stack and curr_height < heights[stack[-1]]:
                h = heights[stack.pop()]
                w = i if not stack else (i - stack[-1] - 1)
                max_area = max(max_area, h * w)

            stack.append(i)

        return max_area

File: 0084-Largest-Rectangle-in-Histogram/solution.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        int n = heights.size();
        stack<int> st; // Stores indices with strictly increasing heights
        int max_area = 0;
        
        for (int i = 0; i <= n; ++i) {
            // Use 0 as a sentinel height at index n to flush all remaining elements
            int curr_height = (i == n) ? 0 : heights[i];
            
            while (!st.empty() && curr_height < heights[st.top()]) {
                int h = heights[st.top()];
                st.pop();
                
                // If stack is empty, width extends all the way back to index 0
                int w = st.empty() ? i : (i - st.top() - 1);
                max_area = max(max_area, h * w);
            }
            
            st.push(i);
        }
        
        return max_area;
    }
};
